FLEXIBLE_INDEXES_NOTES.md

```diff
@@ -0,0 +1,396 @@
+
+In the example above a coordinate is also created for the `x` dimension:
+
+```python
+>>> da.x
+<xarray.DataArray 'x' (x: 2)>
+array([('a', 0), ('b', 1)], dtype=object)
+Coordinates:
+  * x        (x) MultiIndex
+  - lvl1     (x) object 'a' 'b'
+  - lvl2     (x) int64 0 1
+```
+
+With the new proposed data model, this wouldn't be a requirement anymore: there is no concept of a dimension-index. However, some users might still rely on the `x` coordinate so we could still (temporarily) support it for backwards compatibility.
+
+Besides `pandas.MultiIndex`, there may be other situations where we would like to reuse an existing index in a new Dataset/DataArray (e.g., when the index is very expensive to build), and which might require implicit creation of one or more coordinates.
+
+The example given here is quite confusing, though: this is not an easily predictable behavior. We could entirely avoid the implicit creation of coordinates, e.g., using a helper function that generates coordinate + index dictionaries that we could then pass directly to the DataArray/Dataset constructor:
+
+```python
+>>> coords_dict, index_dict = create_coords_from_index(midx, dims='x', include_dim_coord=True)
+>>> coords_dict
+{'x': <xarray.Variable (x: 2)>
+ array([('a', 0), ('b', 1)], dtype=object),
+ 'lvl1': <xarray.Variable (x: 2)>
+ array(['a', 'b'], dtype=object),
+ 'lvl2': <xarray.Variable (x: 2)>
+ array([0, 1])}
+>>> index_dict
+{('lvl1', 'lvl2'): midx}
+>>> xr.DataArray([1.0, 2.0], dims='x', coords=coords_dict, indexes=index_dict)
+<xarray.DataArray (x: 2)>
+array([1., 2.])
+Coordinates:
+    x        (x) object ('a', 0) ('b', 1)
+  * lvl1     (x) object 'a' 'b'
+  * lvl2     (x) int64 0 1
+```
+
+### 2.2.5 Immutable indexes
+
+Some underlying indexes might be mutable (e.g., a tree-based index structure that allows dynamic addition of data points) while other indexes like `pandas.Index` aren't. To keep things simple, it is probably better to continue considering all indexes in Xarray as immutable (as well as their corresponding coordinates, see [Section 2.4.1](#241-mutable-coordinates)).
+
+### 2.3 Index access
+
+#### 2.3.1 Dataset/DataArray's `indexes` property
+
+The `indexes` property would allow easy access to all the indexes used in a Dataset/DataArray. There may be different options for the type returned:
+
+A. `Dict[CoordinateName, XarrayIndex]`: keys are coordinate names and values may be duplicated
+B. `Dict[CoordinateNames, XarrayIndex]`: keys may represent one or more coordinate names and values are unique
+
+Option A allows easy index look-up by coordinate name, while option B allows easy iteration through all the indexes. Option A may be more useful than option B for many tasks and may also be less ambiguous if more complex hashable types than `str` are used for `CoordinateName`.
```

Comment:
I think I mentioned this before, but I think option A is much better.

Here's an example implementation of the two "hard" operations with each data format:

Iterate over all indexes with A:
```python
def iter_all(indexes):
    output_ids = set()
    for index in indexes.values():
        index_id = id(index)
        if index_id not in output_ids:
            output_ids.add(index_id)
            yield index
```

Look-up by coordinate name with option B (basically needs to build up the "option A" representation internally):
```python
def lookup_by_name(indexes, coord_name):
    coord_name_to_index = {}
    for names, index in indexes.items():
        # this would be even worse if we needed to check whether
        # names is a tuple or a single name
        for name in names:
            coord_name_to_index[name] = index
    return coord_name_to_index[coord_name]
```

Comparing the run-time by computational complexity:

- Look-up by coordinate name: `O(1)` with option A vs `O(number_of_indexed_coordinates)` with Option B
- Iterate over all unique indexes: `O(number_of_indexed_coordinates)` with option A vs `O(number_of_unique_indexes)`

I expect "look-up by coordinate name" would be most common in performance limited situation (e.g., in indexing operations), and there's a much more meaningful speed-up there. In most cases, I would guess `number_of_indexed_coordinates` is not much larger than `number_of_unique_indexes`.

There's also the whole separate issue of option B potentially allowing both strings and tuples of strings as keys. Even if we do option B, we should only use tuples to avoid potential ambiguity and needing to handle both cases every time the index is used.

Reply:
Agreed, you already convinced me before that option A is better (not sure that performance is really critical here for only a few coordinates / indexes per dataset, but in general I agree with your arguments in favor of A).

I intentionally left the options in the notes (along with some explanation of why one an option is better than the other), but if we all agree then I can remove option B here.